### ISSUE TITLE
[Bug] Esbuild removes unused imports, despite tree-shaking disabled

### ISSUE THREAD

**Issue by brillout** [evanw/esbuild#3034]
For JavaScript, esbuild keeps the import which is expected: https://hyrious.me/esbuild-repl/?version=0.17.14&mode=build&modules=%5B%5B%22main.js%22%2C%22import+%7B+SomeImport+%7D+from+%27.%2Fsome-file%27%5Cn%22%2C1%5D%5D&buildOptions=%7B%22bundle%22%3Afalse%2C%22format%22%3A%22esm%22%2C%22splitting%22%3Afalse%2C%22treeShaking%22%3Afalse%2C%22minify%22%3Afalse%2C%22keepNames%22%3Atrue%2C%22sourcesContent%22%3Afalse%2C%22ignoreAnnotations%22%3Atrue%2C%22charset%22%3A%22utf8%22%2C%22sourcemap%22%3Afalse%2C%22platform%22%3A%22neutral%22%7D

For TypeScript, esbuild removes the import which is erroneous: https://hyrious.me/esbuild-repl/?version=0.17.14&mode=build&modules=%5B%5B%22main.ts%22%2C%22import+%7B+SomeImport+%7D+from+%27.%2Fsome-file%27%5Cn%22%2C1%5D%5D&buildOptions=%7B%22bundle%22%3Afalse%2C%22format%22%3A%22esm%22%2C%22splitting%22%3Afalse%2C%22treeShaking%22%3Afalse%2C%22minify%22%3Afalse%2C%22keepNames%22%3Atrue%2C%22sourcesContent%22%3Afalse%2C%22ignoreAnnotations%22%3Atrue%2C%22charset%22%3A%22utf8%22%2C%22sourcemap%22%3Afalse%2C%22platform%22%3A%22neutral%22%7D.

This breaks the code's semantics as the imported module may have side effects.

The import is removed even if tree-shaking is disabled.

**Comment by evanw** [comment]
That's how TypeScript itself works: https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgZQiApgSXNeBfOAMyjTgHIA6AegGc10BaQ4AG3TICgg. I'm closing this issue as invalid because esbuild is working as intended.

TypeScript has at least three separate ways of changing this, and esbuild currently supports two of them: [`preserveValueImports`](https://www.typescriptlang.org/tsconfig#preserveValueImports) and [`importsNotUsedAsValues`](https://www.typescriptlang.org/tsconfig#importsNotUsedAsValues). I'm planning on adding support for the third way soon ([`verbatimModuleSyntax`](https://www.typescriptlang.org/tsconfig#verbatimModuleSyntax) was added to TypeScript recently in version 5.0). 

TL;DR you have to configure TypeScript to preserve import statements regardless of whether you use esbuild or tsc to transpile it.

**Comment by brillout** [comment]
Thanks the explanation.

It's surprising that TypeScript breaks side-effect semantics even though it's very strict about preserving semantics in other areas.
